File: leetcode/110_balancedbinarytree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        def getDepth(root, depth):
            if root is None:
                return depth
            l = getDepth(root.left, depth+1)
            r = getDepth(root.right, depth+1)
            if l == -1 or r == -1 or abs(l-r) > 1:
                return -1
            return max(l,r)
        return getDepth(root, 1) != -1

File: leetcode/110_balancedbinarytree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
    if (!root) return true
    const res = goDown(root, 1)
    if (typeof res === 'boolean') return false
    return true
};

function goDown(root: TreeNode | null, currDepth: number): number | boolean {
    if (!root) return currDepth - 1
    const leftDepth = goDown(root.left, currDepth + 1)
    const rightDepth = goDown(root.right, currDepth + 1)
    if (typeof leftDepth === 'number' && typeof rightDepth === 'number') {
        if (Math.abs(leftDepth - rightDepth) > 1) return false
        return Math.max(leftDepth, rightDepth)
    } else {
        return false
    }
    
}
